### PR TITLE
refactor(Core/SmartScripts): Change invoker to be WorldObject

### PR DESCRIPTION
**Pull request by Kitzunu**
## Changes Proposed:

This PR proposes changes to:
-  [x] Core (units, players, creatures, game systems).

## Tests Performed:

This PR has been:

I have only tested with https://github.com/azerothcore/azerothcore-wotlk/pull/19375

I have not tested if this breaks anything

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [x] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

Check SAI with invoker, so it still works

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 1]
```diff
@@ -232,7 +239,7 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
                     sCreatureTextMgr->SendChat(target->ToCreature(), uint8(e.action.simpleTalk.textGroupID), IsPlayer(GetLastInvoker()) ? GetLastInvoker() : 0);
                 else if (IsPlayer(target) && me)
                 {
-                    Unit* templastInvoker = GetLastInvoker();
+                    Unit* templastInvoker = GetLastInvoker()->ToUnit();
```
Comment (PkllonG) [inline src/server/game/AI/SmartScripts/SmartScript.cpp:242]: GetLastInvoker() == nullptr
Server Crash

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 2]
```diff
@@ -783,7 +790,7 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
         case SMART_ACTION_INVOKER_CAST:
         {
             // Can be used for area trigger cast
-            Unit* tempLastInvoker = GetLastInvoker(unit);
+            Unit* tempLastInvoker = GetLastInvoker(unit)->ToUnit();
```
Comment (PkllonG) [inline src/server/game/AI/SmartScripts/SmartScript.cpp:793]: Same as before

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 3]
```diff
@@ -3399,8 +3416,8 @@ void SmartScript::GetTargets(ObjectVector& targets, SmartScriptHolder const& e,
                 targets.push_back(scriptTrigger);
             break;
         case SMART_TARGET_ACTION_INVOKER_VEHICLE:
-            if (scriptTrigger && scriptTrigger->GetVehicle() && scriptTrigger->GetVehicle()->GetBase())
-                targets.push_back(scriptTrigger->GetVehicle()->GetBase());
+            if (scriptTrigger && scriptTrigger->ToUnit()->GetVehicle() && scriptTrigger->ToUnit()->GetVehicle()->GetBase())
+                targets.push_back(scriptTrigger->ToUnit()->GetVehicle()->GetBase());
```
Comment (PkllonG) [inline src/server/game/AI/SmartScripts/SmartScript.cpp:3420]: scriptTrigger->ToUnit() == nullptr
Server Crash

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 4]
```diff
@@ -4356,7 +4373,7 @@ void SmartScript::ProcessEvent(SmartScriptHolder& e, Unit* unit, uint32 var0, ui
                     return;
                 }
 
-                ProcessAction(e, GetLastInvoker());
+                ProcessAction(e, GetLastInvoker()->ToUnit());
```
Comment (PkllonG) [inline src/server/game/AI/SmartScripts/SmartScript.cpp:4376]: Same as before

**Review by Nyeriah — APPROVED** [review]

**Comment by avarishd** [comment]
ATs setting data do NOT work https://discord.com/channels/217589275766685707/1110284108594155573/1263113557961211967

target_type 7 only
I have tested creatures with event_type 4,61,62,8,54,105,74,70,64,16,17,14.
Tested GOs with event_type 61,62,64,70

<details>



https://github.com/user-attachments/assets/feb18174-e460-4852-aaa4-e7760abd19f8


https://github.com/user-attachments/assets/957822e9-fbf2-493d-9060-4dd96dba2481


https://github.com/user-attachments/assets/41a5072f-4eaa-487a-bdd5-21363fe891fc


https://github.com/user-attachments/assets/a9630f0b-60c0-44ea-bf6b-36e839d28462


https://github.com/user-attachments/assets/b567b290-daca-41dc-8a46-026ee5200869


https://github.com/user-attachments/assets/beb8d512-4d40-40b4-86be-3fb2f29f12ef


https://github.com/user-attachments/assets/8902b11b-aa54-4bec-8128-0108ec684500

b437c60a00b1c91b2cc508fc2286afe8

https://github.com/user-attachments/assets/45b72f20-c989-4efc-9ba8-52c630804c76


https://github.com/user-attachments/assets/f2a9ebc4-38d4-4e61-8ae1-49bd9e75e97a


https://github.com/user-attachments/assets/925474b1-9ea2-40d4-ad0d-42ed3f70b6de

</details>

**Comment by avarishd** [comment]
ATs seem to have been fixed.

<details>

https://github.com/user-attachments/assets/806d2f80-67c0-454d-ad96-722977d253a4

https://github.com/user-attachments/assets/0ddd957c-549c-4ee4-9e21-b4339e45967c

</details>

**Review by Nyeriah — APPROVED** [review]

**Pull request merged**